src/main/scala/loamstream/drm/JobMonitor.scala

```diff
@@ -75,9 +75,9 @@ final class JobMonitor(
     //TODO: find a better way to stop or shut down `ticks` :\
     val keepPolling: ValueBox[Boolean] = ValueBox(true)
     
-    val ticks = Observable.interval(period, scheduler).takeUntil(stopSignal).share
+    val ticks = Observable.interval(period, scheduler).onBackpressureDrop.takeUntil(stopSignal).share
     
-    def poll(): Map[String, Try[DrmStatus]] = poller.poll(jobIds)
+    def poll(): Map[DrmTaskId, Try[DrmStatus]] = poller.poll(jobIds)
```

Comment:
It would be helpful (for me in the PR) to have a rough overview of this JobMonitor a bit. It's given a set of jobs to monitor and has a polling frequency. I assume this is all happening in another thread, but it's hard to tell.

```
    val pollResults = ticks
      .takeWhile(_ => shouldContinue)
      .map(_ => poll())
      .until(allFinished(keepPolling))
      .replay
```

Appears to be the key code. But, it's difficult to get a sense of what an execution profile would look like for it. For example, it could be one of the following (and other possibilities, too):

1. Wait dt (by freq) seconds, then poll all the tasks being monitored, when all are done, stop monitoring, otherwise loop again.

2. Wait dt seconds, then poll the next task being monitored (just one), repeat (with waits) until all tasks have been checked, stop if all done.

I ask about this because (#1) if `dt` is too low and the number of tasks being monitored is high, then it's possible there's never a delay between checks due to how long it takes to check the status of each job (is each check a separate call?). 

But, if using (#2) then if `dt` is too high, you could end up with a situation where there are 10,000 tasks and a `dt` of even 0.1 seconds means up to 15 minutes could go by between when all the tasks are done and the JobMonitor actually reports that they are done.

Maybe (if one of the above) due to use-cases, neither of the above are concerns. Or maybe it's something else entirely different that would be good to understand/have commented?

Reply:
I'm glad you noticed this.  `JobMonitor` is complex, partly out of necessity, partly not.  What's actually happening is closest to #1.  

This is tricky, since it's not possible to speed up polling _that_ much without a bigger refactoring, and maybe not even then.  DRMAA, the library LS uses to talk to Uger, is single threaded.  You can't ask for the status of multiple jobs at once; you have to do it in sequence, which is what `DrmaaPoller` does.  One way around this might be to run the `qstat` command-line tool and parse its output to determine job status (which is basiclly what happens at EBI on their LSF cluster).  That queries the same underlying UGER-side data structures as DRMAA does, so throughput might be constrained by the same underlying lock in any case.  

It's worth investigating though, since the scenario you mentioned (too-low dt with large numbers of jobs) is likely to happen now.

And, since polling the status of multiple jobs isn't (and can't currently) happen concurrently, we could easily run into the scenario you raised for #2: Since polling happens on-job-at-a-time, even if it's done asynchronously for chunks of jobs, given 10k jobs (very feasible now) polling all of them could take minutes.  This could be mitigated by having `Poller`s produce `Observable[(DrmTaskId, Try[DrmStatus])]`s instead of `Observable[Map[DrmTaskId, Try[DrmStatus]]`s - that is, emit the polling result for each job as it's available instead of waiting for all of them.

Reply:
Having dealt with something similar in the past, I would recommend doing option #2, but in a slightly different algorithm (that isn't well suited to nice FP):

1. You have a list of enumerated jobs to track: `Vector((Int, Job))`
2. Wait dt seconds.
3. Poll job at head of `val i, job = trackList.last`.
4. If done, pop from list, mark status of `trackStatus(i)` and goto 3.
5. Rotate list (move last to first and shift all) and goto 2.

Probably a significant change, but works quite effectively. The shifting is annoying, but prevents instances where the first job being tracked takes the longest and you jump from 0% done to 100% done instantly, which can also cause problems if the status check is slow or rate limited, causing issues with the pop-and-repeat.

Reply:
These are very good points, but too big for this already huge PR, IMO.  This is definitely something that will need revisited, since the current code won't handle polling large numbers (10k) jobs well.  I settled for mitigating the issue by emitting polling results asynchronously, one at a time, instead of waiting for the results for all N jobs and emitting them at once.

A fundamental problem is that polling Uger jobs can't happen concurrently, but that doesn't mean we need to wait for all of them.

`JobMonitor` in particular has evolved through many refactorings and requirement changes, and has become quite complex.  It's probably do for a re-write, but it's of critical importance, and it "works", so it's never made sense to change it before now.